backend/dataall/modules/dataset_sharing/services/data_sharing_service.py

```diff
@@ -226,3 +249,106 @@ def revoke_share(cls, engine: Engine, share_uri: str):
+            dataset_lock = (
+                session.query(DatasetLock)
+                .filter(
+                    and_(
+                        DatasetLock.datasetUri == dataset_uri,
+                        ~DatasetLock.isLocked
+                    )
+                )
+                .first()
+            )
+
+            # Check if dataset_lock is not None before attempting to update
+            if dataset_lock:
+                # Update the attributes of the DatasetLock object
+                dataset_lock.isLocked = True
+                dataset_lock.acquiredBy = share_uri
+
+                session.commit()
+                return True
+            else:
+                log.info("DatasetLock not found for the given criteria.")
+                return False
+
+        except Exception as e:
+            session.rollback()
+            log.error("Error occurred while acquiring lock:", e)
+            return False
+
+    @staticmethod
+    def acquire_lock_with_retry(dataset_uri, session, share_uri):
+        for attempt in range(MAX_RETRIES):
+            try:
+                log.info(f"Attempting to acquire lock for dataset {dataset_uri} by share {share_uri}...")
+                lock_acquired = DataSharingService.acquire_lock(dataset_uri, session, share_uri)
+                if lock_acquired:
+                    return True
+
+                log.info(
+                    f"Lock for dataset {dataset_uri} already acquired. Retrying in {RETRY_INTERVAL} seconds...")
+                sleep(RETRY_INTERVAL)
+
+            except Exception as e:
+                log.error("Error occurred while retrying acquiring lock:", e)
+                return False
+
+        log.info(f"Max retries reached. Failed to acquire lock for dataset {dataset_uri}")
+        return False
+
+    @staticmethod
+    def release_lock(dataset_uri, session, share_uri):
+        """
+        Releases the lock on the dataset identified by dataset_uri.
+
+        Args:
+            dataset_uri: The ID of the dataset for which the lock is being released.
+            session (sqlalchemy.orm.Session): The SQLAlchemy session object used for interacting with the database.
+            share_uri: The ID of the share that is attempting to release the lock.
+
+        Returns:
+            bool: True if the lock is successfully released, False otherwise.
+        """
+        try:
+            log.info(f"Releasing lock for dataset: {dataset_uri} last acquired by share: {share_uri}")
+            query = (
+                session.query(DatasetLock)
+                .filter(
+                    and_(
+                        DatasetLock.datasetUri == dataset_uri,
+                        DatasetLock.isLocked == True,
+                        DatasetLock.acquiredBy == share_uri
+                    )
+                )
+            )
+
+            query.update(
+                {
+                    "isLocked": False,
+                    "acquiredBy": share_uri
```

Comment:
do we need to update  `"acquiredBy": share_uri` - is it not already this share_uri and we can just update `isLocked`. If anything I think it makes most sense to update `"acquiredBy": ''` if the lock is being released (only if `acquiredBy` can be nullable

Reply:
The acquiredBy column was more to indicate lock last acquired by which share. 
However, I can make it nullable and reset it to '' when we release lock too. 
I just thought knowing the last share that acquired it, might be useful in debugging. 

I have pushed a revision where acquiredBy can be null and acquiredBy = '' while releasing lock.